<a href="https://colab.research.google.com/github/sarath-777/CodeGenerationFromGithubFiles/blob/main/DependencyGraph_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This gives the list of all groq api models available

In [ ]:
import requests

from google.colab import userdata

GROQ_API_KEY = userdata.get('groqapi')

response = requests.get(
    "https://api.groq.com/openai/v1/models",
    headers={"Authorization": f"Bearer {GROQ_API_KEY}"}
)

print(response.json())  # Print available models


{'object': 'list', 'data': [{'id': 'qwen-qwq-32b', 'object': 'model', 'created': 1741214760, 'owned_by': 'Alibaba Cloud', 'active': True, 'context_window': 131072, 'public_apps': None}, {'id': 'distil-whisper-large-v3-en', 'object': 'model', 'created': 1693721698, 'owned_by': 'Hugging Face', 'active': True, 'context_window': 448, 'public_apps': None}, {'id': 'llama-3.2-90b-vision-preview', 'object': 'model', 'created': 1727226914, 'owned_by': 'Meta', 'active': True, 'context_window': 8192, 'public_apps': None}, {'id': 'allam-2-7b', 'object': 'model', 'created': 1737672203, 'owned_by': 'SDAIA', 'active': True, 'context_window': 4096, 'public_apps': None}, {'id': 'llama-3.3-70b-versatile', 'object': 'model', 'created': 1733447754, 'owned_by': 'Meta', 'active': True, 'context_window': 32768, 'public_apps': None}, {'id': 'llama3-70b-8192', 'object': 'model', 'created': 1693721698, 'owned_by': 'Meta', 'active': True, 'context_window': 8192, 'public_apps': None}, {'id': 'llama-guard-3-8b', '

In [ ]:
pip install networkx requests PyGithub


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.5/410.5 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 856.7/856.7 kB 20.8 MB/s eta 0:00:00


In [ ]:
pip install PyGithub networkx


2️⃣ fetch_repo.py → Clone & Extract Dependencies from GitHub
This script:
✅ Fetches the GitHub repository
✅ Extracts dependencies (imports, function calls, API routes, database changes)
✅ Builds a dependency graph

In [ ]:
import os
import re
import json
import networkx as nx
from github import Github, InputGitTreeElement

# GitHub Authentication
from google.colab import userdata
GITHUB_TOKEN = userdata.get('githubReadAndWrite')
REPO_NAME = "sarath-777/demo_angular_springboot"  # Change this to your GitHub repo

# Initialize GitHub API client
g = Github(GITHUB_TOKEN)
repo = g.get_repo(REPO_NAME)

# Create a directed graph (dependencies have direction)
graph = nx.DiGraph()

def fetch_repo_files(repo):
    """Fetch all file paths in the repository."""
    contents = repo.get_contents("")
    files = []

    while contents:
        file_content = contents.pop(0)
        if file_content.type == "dir":
            contents.extend(repo.get_contents(file_content.path))
        else:
            files.append(file_content.path)

    return files

def extract_dependencies(file_content, file_path):
    """Extract dependencies from Angular & Spring Boot files."""
    dependencies = []

    # Detect Angular imports (TypeScript)
    if file_path.endswith(".ts"):
        matches = re.findall(r'import .* from ["\'](.*)["\'];', file_content)
        dependencies.extend(matches)

    # Detect Java imports
    elif file_path.endswith(".java"):
        matches = re.findall(r'import (.*);', file_content)
        dependencies.extend(matches)

    # Detect API calls
    api_matches = re.findall(r'httpClient\.(get|post|put|delete)\(["\'](.*?)["\']', file_content)
    dependencies.extend([api[1] for api in api_matches])

    return dependencies

def build_dependency_graph(repo):
    """Build a graph linking files based on dependencies."""
    files = fetch_repo_files(repo)

    for file_path in files:
        # Process only code files
        if not file_path.endswith((".ts", ".java", ".json", ".html")):
            continue  # Skip non-text files

        try:
            file_content = repo.get_contents(file_path).decoded_content.decode("utf-8")
        except UnicodeDecodeError:
            print(f"⚠️ Skipping binary/non-UTF-8 file: {file_path}")
            continue  # Skip files that can't be decoded as UTF-8

        graph.add_node(file_path)  # Add file as a node

        dependencies = extract_dependencies(file_content, file_path)

        for dep in dependencies:
            graph.add_edge(file_path, dep)  # Create dependency link

    return graph

# Build the graph
dependency_graph = build_dependency_graph(repo)

# Save graph locally in Colab
GRAPH_FILE = "dependency_graph.gml"
nx.write_gml(dependency_graph, GRAPH_FILE)
print(f"Dependency graph saved locally as '{GRAPH_FILE}'")

# Upload to GitHub in the `depGraphs/` folder
def upload_to_github(local_file, repo, github_path):
    """Upload the local file to GitHub repo."""
    with open(local_file, "r") as file:
        content = file.read()

    try:
        # Check if file already exists in GitHub
        contents = repo.get_contents(github_path)
        repo.update_file(github_path, "Updated dependency graph", content, contents.sha)
        print(f"✅ Updated '{github_path}' in GitHub repo.")
    except:
        repo.create_file(github_path, "Added dependency graph", content)
        print(f"✅ Uploaded '{github_path}' to GitHub repo.")

# Upload dependency graph to GitHub
upload_to_github(GRAPH_FILE, repo, "depGraphs/dependency_graph.gml")


Dependency graph saved locally as 'dependency_graph.gml'
✅ Uploaded 'depGraphs/dependency_graph.gml' to GitHub repo.


convert dependency graph to json format

In [ ]:
import networkx as nx
import json

GRAPH_FILE = "dependency_graph.gml"
OUTPUT_JSON_FILE = "dependency_graph.json"

# Load the dependency graph
dependency_graph = nx.read_gml(GRAPH_FILE)

# Convert to JSON
graph_json = {
    "nodes": list(dependency_graph.nodes),
    "edges": list(dependency_graph.edges)
}

# Save as JSON
with open(OUTPUT_JSON_FILE, "w") as f:
    json.dump(graph_json, f, indent=4)

print(f"✅ Dependency graph saved as JSON: {OUTPUT_JSON_FILE}")


✅ Dependency graph saved as JSON: dependency_graph.json


code to send to groq api


In [ ]:
import json
import requests


from google.colab import userdata

GROQ_API_KEY = userdata.get('groqapi')  # Replace with your Groq API key
GROQ_MODEL = "llama3-70b"  # Model name (choose an appropriate one)

# Load the dependency graph JSON
with open("dependency_graph.json", "r") as f:
    dependency_graph_json = json.load(f)

# New feature request (user input)
new_feature_request = """
I want to add a new feature that allows users to reset their password.
This requires:
- A new API endpoint in the backend
- A new frontend UI in Angular
- Updating existing authentication logic
"""

# Construct LLM Prompt
prompt = f"""
You are an AI assistant helping in software development.
Given the following dependency graph of a GitHub project and a new feature request,
identify which files will be affected and need modifications.

Dependency Graph:
{json.dumps(dependency_graph_json, indent=4)}

New Feature Request:
{new_feature_request}

Please analyze and list:
1. Files that need to be created or modified.
2. Changes required in each file.
"""

# Call Groq API
response = requests.post(
    "https://api.groq.com/v1/chat/completions",
    headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
    json={
        "model": GROQ_MODEL,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.5
    }
)

# Process Response
llm_output = response.json()
suggestions = llm_output.get("choices", [{}])[0].get("message", {}).get("content", "")

print("📌 Suggested Changes:\n", suggestions)


📌 Suggested Changes:
 


the order of execution should be:
-llm fetches github repo files and structure
- llm analyzes the github repo files and structure
- llm analyzes the dependecy graph
- user gives new feature requeest
- llm  with the knowledge of github repo gives the code to the new feature request
- now based on the knowlege from the dependecy graph, the model identifies which files will be affected by the new feature suggestions and llm suggests to make modifications to those affected files inorder to avoid any errors or conflicts

Execution Steps:
1️⃣ LLM Fetches GitHub Repo Files & Structure
2️⃣ LLM Analyzes the GitHub Repo (Code, APIs, Components, DB, etc.)
3️⃣ LLM Analyzes the Dependency Graph
4️⃣ User Gives a New Feature Request
5️⃣ LLM Generates New Feature Code (Backend, Frontend, DB, API Updates)
6️⃣ LLM Identifies Affected Files & Suggests Modifications

1️⃣ & 2️⃣ LLM Fetches & Analyzes GitHub Repo

Modify the existing fetch_repo.py to send data directly to Groq API.

In [ ]:
import json
import requests
from github import Github

# GitHub & Groq API credentials
from google.colab import userdata

GITHUB_TOKEN = userdata.get('githubReadAndWrite')
GROQ_API_KEY = userdata.get('groqapi')
REPO_NAME = "sarath-777/demo_angular_springboot"
GROQ_MODEL = "llama3-70b-8192"

# Initialize GitHub client
g = Github(GITHUB_TOKEN)
repo = g.get_repo(REPO_NAME)

def fetch_repo_structure(repo):
    """Fetch all files and their contents from the GitHub repo."""
    contents = repo.get_contents("")
    repo_data = {}

    while contents:
        file_content = contents.pop(0)
        if file_content.type == "dir":
            contents.extend(repo.get_contents(file_content.path))
        else:
            try:
                content = file_content.decoded_content.decode("utf-8")
            except UnicodeDecodeError:
                content = None  # Ignore non-text files
            repo_data[file_content.path] = content

    return repo_data

# Fetch repo data
repo_structure = fetch_repo_structure(repo)

# Send repo data to Groq API for analysis
prompt = f"""
Analyze the following GitHub repository structure.
Extract important files, their purposes, API endpoints, components, and database interactions.

Repository Data:
{json.dumps(repo_structure, indent=4)}
"""

response = requests.post(
    "https://api.groq.com/openai/v1/chat/completions",
    headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
    json={
        "model": GROQ_MODEL,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.3
    }
)

# Process Response

response_json = response.json()  # Get JSON response
print("🚀 Full API Response:\n", json.dumps(response_json, indent=4))  # Debug

if "choices" in response_json:
    llm_analysis = response_json["choices"][0]["message"]["content"]
    print("📌 LLM Analysis of GitHub Repo:\n", llm_analysis)
else:
    print("❌ Error: 'choices' key missing in API response!")

llm_analysis = response.json()["choices"][0]["message"]["content"]
print("📌 LLM Analysis of GitHub Repo:\n", llm_analysis)


🚀 Full API Response:
 {
    "error": {
        "message": "Request too large for model `llama3-70b-8192` in organization `org_01jn0nr2n4fmt855dgkax3cb6b` service tier `on_demand` on tokens per minute (TPM): Limit 6000, Requested 168647, please reduce your message size and try again. Visit https://console.groq.com/docs/rate-limits for more information.",
        "type": "tokens",
        "code": "rate_limit_exceeded"
    }
}
❌ Error: 'choices' key missing in API response!


KeyError: 'choices'

🔥 Issue: Request Too Large for Groq API
Your request exceeds the 6,000 token per minute (TPM) limit on the Groq API. The error message states that your request used 168,647 tokens, which is too large.

In [ ]:
import json
import requests
from github import Github
from google.colab import userdata

# GitHub & Groq API credentials
GITHUB_TOKEN = userdata.get('githubReadAndWrite')
GROQ_API_KEY = userdata.get('groqapi')
REPO_NAME = "sarath-777/demo_angular_springboot"
GROQ_MODEL = "llama3-70b-8192"

# Initialize GitHub client
g = Github(GITHUB_TOKEN)
repo = g.get_repo(REPO_NAME)

def fetch_repo_structure(repo, max_file_size=10000):
    """Fetch selected files (skip large/minor files) from the GitHub repo."""
    contents = repo.get_contents("")
    repo_data = {}

    while contents:
        file_content = contents.pop(0)
        if file_content.type == "dir":
            contents.extend(repo.get_contents(file_content.path))
        else:
            # Skip binary files and large files
            if file_content.size > max_file_size or not file_content.path.endswith((".ts", ".java", ".html", ".json")):
                continue
            try:
                content = file_content.decoded_content.decode("utf-8")
                # Summarize content if file is large
                if len(content) > 3000:
                    content = f"Summarized: {content[:1500]} ... {content[-1500:]}"
            except UnicodeDecodeError:
                content = None  # Ignore non-text files

            repo_data[file_content.path] = content

    return repo_data

# Fetch reduced repo data
repo_structure = fetch_repo_structure(repo)

# Send repo data to Groq API in **smaller batches**
chunk_size = 1000  # Number of characters per batch
repo_json = json.dumps(repo_structure, indent=4)

for i in range(0, len(repo_json), chunk_size):
    batch = repo_json[i:i+chunk_size]

    prompt = f"""
    Analyze the following GitHub repository files:
    {batch}
    Extract APIs, frontend components, and database details.
    """

    response = requests.post(
        "https://api.groq.com/openai/v1/chat/completions",
        headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
        json={
            "model": GROQ_MODEL,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": 0.3
        }
    )

    # Process Response
    response_json = response.json()

    if "choices" in response_json:
        llm_analysis = response_json["choices"][0]["message"]["content"]
        print(f"📌 LLM Analysis (Batch {i // chunk_size + 1}):\n", llm_analysis)
    else:
        print(f"❌ Error in batch {i // chunk_size + 1}: 'choices' key missing in API response!")


📌 LLM Analysis (Batch 1):
 Based on the provided `.eslintrc.json` file, here's what can be extracted:

**APIs:**

* None explicitly mentioned in this file. However, since it's an Angular project, it's likely that APIs will be defined in separate files, such as `environment.ts` or `api.service.ts`.

**Frontend Components:**

* The file mentions Angular-specific ESLint rules, which suggests that the frontend is built using Angular.
* The `directive-selector` and `component-selector` rules imply that the project uses Angular directives and components.
* The `prefix` property in these rules is set to `"app"`, which might indicate that the application prefix is "app" (e.g., `app-my-component`).

**Database Details:**

* None mentioned in this file. The `.eslintrc.json` file is primarily concerned with ESLint configuration for the frontend code, and database details are typically stored in separate configuration files or environment variables.

Other observations:

* The project uses TypeScr

3️⃣ LLM Analyzes Dependency Graph

In [ ]:
import networkx as nx

# Load the dependency graph
dependency_graph = nx.read_gml("dependency_graph.gml")

def convert_graph_to_dict(graph):
    """Convert NetworkX graph to a dictionary format for LLM."""
    graph_dict = {}
    for node in graph.nodes:
        graph_dict[node] = list(graph.successors(node))  # Get dependent files

    return graph_dict

# Convert graph to JSON format
dependency_graph_dict = convert_graph_to_dict(dependency_graph)
dependency_graph_json = json.dumps(dependency_graph_dict, indent=4)

# Send dependency graph data to Groq API in smaller batches
chunk_size = 1000  # Number of characters per batch

for i in range(0, len(dependency_graph_json), chunk_size):
    batch = dependency_graph_json[i:i+chunk_size]

    prompt = f"""
    This is the dependency graph of a GitHub repository:
    {batch}
    - Understand which files depend on which.
    - Identify critical dependencies between frontend (Angular) and backend (Spring Boot).
    """

    response = requests.post(
        "https://api.groq.com/openai/v1/chat/completions",
        headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
        json={
            "model": GROQ_MODEL,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": 0.3
        }
    )

    # Process Response
    response_json = response.json()

    if "choices" in response_json:
        llm_dependency_analysis = response_json["choices"][0]["message"]["content"]
        print(f"📌 LLM Dependency Analysis (Batch {i // chunk_size + 1}):\n", llm_dependency_analysis)
    else:
        print(f"❌ Error in batch {i // chunk_size + 1}: 'choices' key missing in API response!")


FileNotFoundError: [Errno 2] No such file or directory: 'dependency_graph.gml'

21/03/2025

Instead of manually creating the dependency graph send the github file structure to the groq api model

1️⃣ Create the Dependency Graph using the LLM
Instead of manually parsing file imports and API calls, let's send a request to the Groq model to analyze the repo and generate a dependency graph.

Code to Generate Dependency Graph via Groq API:
python
Copy
Edit


In [ ]:
!pip install PyGithub


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.5/410.5 kB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 856.7/856.7 kB 32.0 MB/s eta 0:00:00


In [ ]:
import json
import requests
import networkx as nx
import time
import re
from github import Github
from google.colab import userdata

# GitHub & Groq API credentials
GITHUB_TOKEN = userdata.get('githubReadAndWrite')
GROQ_API_KEY = userdata.get('groqapi')
REPO_NAME = "sarath-777/demo_angular_springboot"
GROQ_MODEL = "llama3-70b-8192"

# Initialize GitHub client
g = Github(GITHUB_TOKEN)
repo = g.get_repo(REPO_NAME)
import json
import re

def extract_json(text):
    """Extract JSON data from the LLM response using regex."""
    match = re.search(r'```json\n(.*?)\n```', text, re.DOTALL)
    if match:
        return match.group(1)  # Extract JSON block
    return text  # Use full response if no JSON block is found
def convert_to_networkx(json_data):
    """Convert extracted JSON dependency graph to NetworkX graph."""
    dependency_graph = nx.DiGraph()

    # Ensure correct JSON format
    if "nodes" in json_data and "edges" in json_data:
        for node in json_data["nodes"]:
            dependency_graph.add_node(node["id"], label=node["label"])

        for edge in json_data["edges"]:
            dependency_graph.add_edge(edge["from"], edge["to"], label=edge["label"])
    else:
        raise ValueError("Invalid JSON format for dependency graph!")

    return dependency_graph


def fetch_repo_structure(repo, max_file_size=10000):
    """Fetch selected files (skip large/minor files) from the GitHub repo."""
    contents = repo.get_contents("")
    repo_data = {}

    while contents:
        file_content = contents.pop(0)
        if file_content.type == "dir":
            contents.extend(repo.get_contents(file_content.path))
        else:
            if file_content.size > max_file_size or not file_content.path.endswith((".ts", ".java", ".html", ".json")):
                continue
            try:
                content = file_content.decoded_content.decode("utf-8")
                if len(content) > 3000:
                    content = f"Summarized: {content[:1500]} ... {content[-1500:]}"
            except UnicodeDecodeError:
                content = None

            repo_data[file_content.path] = content

    return repo_data

# Fetch repo data
repo_structure = fetch_repo_structure(repo)

# Convert repo data to JSON
repo_json = json.dumps(repo_structure, indent=4)

# Set chunk size to avoid token limit
chunk_size = 2500  # Reduce batch size to prevent rate limiting

dependency_graph = nx.DiGraph()
all_graph_data = {}

# Function to call Groq API with rate limit handling
def call_groq_api(prompt, retries=3, wait_time=10):
    """Call the Groq API with retry mechanism."""
    for attempt in range(retries):
        response = requests.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
            json={
                "model": GROQ_MODEL,
                "messages": [{"role": "user", "content": prompt}],
                "temperature": 0.3
            }
        )

        response_json = response.json()

        if "choices" in response_json and response_json["choices"]:
            return response_json["choices"][0]["message"]["content"]

        elif 'error' in response_json and 'rate_limit_exceeded' in response_json['error']['code']:
            print(f"⚠️ Rate limit hit. Retrying in {wait_time} seconds...")
            time.sleep(wait_time)
            wait_time *= 2  # Exponential backoff
        else:
            print("❌ API Error:", response_json)
            return None  # Stop retrying on unexpected errors

    return None  # Return None if all retries fail
# Process repo data in chunks
for i in range(0, len(repo_json), chunk_size):
    batch = repo_json[i:i+chunk_size]

    prompt = f"""
    Analyze the following GitHub repository files:
    {batch}
    Identify dependencies between files and create a structured dependency graph.
    Return the graph as a JSON object where each file is a node, and dependencies are edges.
    """

    llm_response = call_groq_api(prompt)

    if llm_response:
        extracted_json = extract_json(llm_response)  # Extract JSON from response

        try:
            parsed_data = json.loads(extracted_json)  # Convert JSON string to dictionary
            batch_graph = convert_to_networkx(parsed_data)  # Convert to NetworkX format
            all_graph_data.update(parsed_data)  # Merge batch data
            print(f"✅ Processed Batch {i // chunk_size + 1}")
        except json.JSONDecodeError:
            print("❌ JSON Decode Error: Could not parse the extracted JSON!")
            print("📌 Raw Extracted JSON:\n", extracted_json)
        except ValueError as e:
            print(f"❌ Graph Conversion Error: {e}")
    else:
        print(f"❌ Failed to process batch {i // chunk_size + 1}")

# Save Graph
graph_path = "/content/dependency_graph.gml"
nx.write_gml(dependency_graph, graph_path)
print(f"✅ Dependency graph saved at: {graph_path}")


❌ JSON Decode Error: Could not parse the extracted JSON!
📌 Raw Extracted JSON:
 After analyzing the provided GitHub repository files, I've identified the dependencies between files and created a structured dependency graph. Here is the graph as a JSON object:

```
{
  "nodes": {
    "angular-with-springboot-main/frontend/todo/.eslintrc.json": {},
    "angular-with-springboot-main/frontend/todo/angular.json": {},
    "angular-with-springboot-main/frontend/todo/src/index.html": {},
    "angular-with-springboot-main/frontend/todo/src/main.ts": {},
    "angular-with-springboot-main/frontend/todo/src/main.server.ts": {},
    "angular-with-springboot-main/frontend/todo/src/styles.css": {},
    "angular-with-springboot-main/frontend/todo/src/favicon.ico": {},
    "angular-with-springboot-main/frontend/todo/src/assets": {},
    "angular-with-springboot-main/frontend/todo/tsconfig.app.json": {},
    "angular-with-springboot-main/frontend/todo/server.ts": {},
    "node_modules/@angular/cli/lib/c

KeyError: 'label'

In [ ]:
with open("dependency_graph.gml", "r") as f:
    print(f.read())  # Prints the graph structure in GML format


FileNotFoundError: [Errno 2] No such file or directory: 'dependency_graph.gml'

In [ ]:
# Print full response for debugging
print("📌 Full API Response:", response_json)

# Check if the API response is valid
if "choices" in response_json and response_json["choices"]:
    try:
        batch_graph_data = json.loads(response_json["choices"][0]["message"]["content"])
        all_graph_data.update(batch_graph_data)  # Merge graph data from all batches
        print(f"✅ Processed Batch {i // chunk_size + 1}")
    except json.JSONDecodeError:
        print("❌ JSON Decode Error: Could not parse the LLM response!")
else:
    print("❌ Error: API response missing 'choices' key or is empty!")


📌 Full API Response: {'id': 'chatcmpl-384833db-d6dd-4dba-91bc-fd5ada55acc1', 'object': 'chat.completion', 'created': 1742531377, 'model': 'llama3-70b-8192', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': "After analyzing the provided GitHub repository files, I've identified the dependencies between files and created a structured dependency graph. Here's the result:\n\n**Dependency Graph:**\n\n```\nApiApplication.java\n  |\n  |---> ApiController.java (uses @SpringBootApplication)\n  |---> MeowFactsController.java (uses @RestController)\n  |---> wheatherValidationApi.java (uses @RestController)\n  |---> ApiServiceImpl.java (uses @Service)\n\nApiController.java\n  |\n  |---> (no dependencies)\n\nMeowFactsController.java\n  |\n  |---> ApiServiceImpl.java (uses @Autowired)\n  |---> ApiService.java (implements interface)\n\nwheatherValidationApi.java\n  |\n  |---> ApiServiceImpl.java (uses @Autowired)\n  |---> ApiService.java (implements interface)\n\nApiService.java\n 

In [ ]:
test_response = requests.post(
    "https://api.groq.com/openai/v1/chat/completions",
    headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
    json={
        "model": GROQ_MODEL,
        "messages": [{"role": "user", "content": "Hello, can you respond?"}],
        "temperature": 0.3
    }
)

print("🔍 Test Response:", test_response.json())


🔍 Test Response: {'id': 'chatcmpl-1516004d-488e-4138-8002-fc163375fb36', 'object': 'chat.completion', 'created': 1742530527, 'model': 'llama3-70b-8192', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': "Hello! Yes, I can respond. I'm happy to chat with you. How are you doing today?"}, 'logprobs': None, 'finish_reason': 'stop'}], 'usage': {'queue_time': 0.20000416899999998, 'prompt_tokens': 16, 'prompt_time': 0.000244708, 'completion_tokens': 23, 'completion_time': 0.076312459, 'total_tokens': 39, 'total_time': 0.076557167}, 'system_fingerprint': 'fp_2e0feca3c9', 'x_groq': {'id': 'req_01jpvekfv2f4hawx547sszfqm1'}}


THIS DESCRIBES HOW TO CREATE DEPENDENCY GRAPH FOR **ANGULAR**

In [ ]:
This output shows the dependency graph of your Angular project as generated by Madge. It lists how different TypeScript files in your src/app directory are interconnected.

Key Observations:
Core Structure:

app.module.ts is importing multiple components (app.component.ts, app.routes.ts, error.component.ts, etc.), making it a central file in the project.

app.routes.ts defines routes for components like error.component.ts, list-todos.component.ts, and login.component.ts.

Component Relationships:

app.component.ts has dependencies on multiple UI components (error.component.ts, footer.component.ts, list-todos.component.ts, etc.).

The login.component.ts, menu.component.ts, and logout.component.ts depend on hardcoded-authentication.service.ts for authentication.

Service Dependencies:

route-gaurd.service.ts depends on hardcoded-authentication.service.ts, indicating that route guards use authentication logic.

Testing Files:

Each component and service has a corresponding .spec.ts file, ensuring unit tests exist for these parts of the application.

Next Steps:
If you're planning to build an LLM-based dependency graph for feature tracking, this data can be used as a reference for constructing an automated metadata structure.

Would you like me to help visualize this dependency graph using a tool or generate a JSON representation?










In [ ]:


C:\Users\287916\Documents\demoAngular_Springboot_project\angular-with-springboot-main\frontend\todo>npx madge --extensions ts src/app
Processed 24 files (1.3s)

app.component.spec.ts
  app.component.ts
app.component.ts
  error/error.component.ts
  footer/footer.component.ts
  list-todos/list-todos.component.ts
  login/login.component.ts
  menu/menu.component.ts
  welcome/welcome.component.ts
app.config.server.ts
  app.config.ts
app.config.ts
  app.routes.ts
app.module.ts
  app.component.ts
  app.routes.ts
  error/error.component.ts
  footer/footer.component.ts
  list-todos/list-todos.component.ts
  login/login.component.ts
  logout/logout.component.ts
  menu/menu.component.ts
  welcome/welcome.component.ts
app.routes.ts
  error/error.component.ts
  list-todos/list-todos.component.ts
  login/login.component.ts
  logout/logout.component.ts
  service/route-gaurd.service.ts
  welcome/welcome.component.ts
error/error.component.spec.ts
  error/error.component.ts
error/error.component.ts
footer/footer.component.spec.ts
  footer/footer.component.ts
footer/footer.component.ts
list-todos/list-todos.component.spec.ts
  list-todos/list-todos.component.ts
list-todos/list-todos.component.ts
login/login.component.spec.ts
  login/login.component.ts
login/login.component.ts
  service/hardcoded-authentication.service.ts
logout/logout.component.spec.ts
  logout/logout.component.ts
logout/logout.component.ts
  service/hardcoded-authentication.service.ts
menu/menu.component.spec.ts
  menu/menu.component.ts
menu/menu.component.ts
  service/hardcoded-authentication.service.ts
service/hardcoded-authentication.service.spec.ts
  service/hardcoded-authentication.service.ts
service/hardcoded-authentication.service.ts
service/route-gaurd.service.spec.ts
  service/route-gaurd.service.ts
service/route-gaurd.service.ts
  service/hardcoded-authentication.service.ts
welcome/welcome.component.spec.ts
  welcome/welcome.component.ts
welcome/welcome.component.ts


C:\Users\287916\Documents\demoAngular_Springboot_project\angular-with-springboot-main\frontend\todo>